pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -61,172 +103,188 @@ type ecrTokenGetter struct {
+
+	credentialprovider.RegisterCredentialProvider("aws-ecr-partition-china",
+		newECRProvider(registryURLTemplateChina, nil))
+}
+
 // Enabled implements DockerConfigProvider.Enabled for the AWS token-based implementation.
 // For now, it gets activated only if AWS was chosen as the cloud provider.
 // TODO: figure how to enable it manually for deployments that are not on AWS but still
 // use ECR somehow?
 func (p *ecrProvider) Enabled() bool {
-	if p.region == "" {
-		klog.Errorf("Called ecrProvider.Enabled() with no region set")
-		return false
-	}
-
-	getter := &ecrTokenGetter{svc: ecr.New(session.New(&aws.Config{
-		Credentials: nil,
-		Region:      &p.region,
-	}))}
-	getter.svc.Handlers.Build.PushFrontNamed(request.NamedHandler{
-		Name: "k8s/user-agent",
-		Fn:   request.MakeAddToUserAgentHandler("kubernetes", version.Get().String()),
-	})
-	getter.svc.Handlers.Sign.PushFrontNamed(request.NamedHandler{
-		Name: "k8s/logger",
-		Fn:   awsHandlerLogger,
-	})
-	p.getter = getter
-
 	return true
 }
 
-// LazyProvide implements DockerConfigProvider.LazyProvide. Should never be called.
-func (p *ecrProvider) LazyProvide() *credentialprovider.DockerConfigEntry {
+// LazyProvide is lazy
+func (p *ecrProvider) LazyProvide(repoToPull string) *credentialprovider.DockerConfigEntry {
 	return nil
 }
 
-// Provide implements DockerConfigProvider.Provide, refreshing ECR tokens on demand
-func (p *ecrProvider) Provide() credentialprovider.DockerConfig {
+// Provide provides credentials from the cache if they are found, or from ECR
+func (p *ecrProvider) Provide(repoToPull string) credentialprovider.DockerConfig {
 	cfg := credentialprovider.DockerConfig{}
 
-	// TODO: fill in RegistryIds?
-	params := &ecr.GetAuthorizationTokenInput{}
-	output, err := p.getter.GetAuthorizationToken(params)
+	parsed, err := splitSchemelessURL(repoToPull)
 	if err != nil {
-		klog.Errorf("while requesting ECR authorization token %v", err)
-		return cfg
-	}
-	if output == nil {
-		klog.Errorf("Got back no ECR token")
+		klog.Errorf("unable to parse repo url %s %v", repoToPull, err)
 		return cfg
 	}
-
-	for _, data := range output.AuthorizationData {
-		if data.ProxyEndpoint != nil &&
-			data.AuthorizationToken != nil {
-			decodedToken, err := base64.StdEncoding.DecodeString(aws.StringValue(data.AuthorizationToken))
-			if err != nil {
-				klog.Errorf("while decoding token for endpoint %v %v", data.ProxyEndpoint, err)
-				return cfg
-			}
-			parts := strings.SplitN(string(decodedToken), ":", 2)
-			user := parts[0]
-			password := parts[1]
-			entry := credentialprovider.DockerConfigEntry{
-				Username: user,
-				Password: password,
-				// ECR doesn't care and Docker is about to obsolete it
-				Email: "not@val.id",
-			}
-
-			klog.V(3).Infof("Adding credentials for user %s in %s", user, p.region)
-			// Add our config entry for this region's registry URLs
-			cfg[p.regionURL] = entry
-
-		}
+	cfg, exists := p.getFromCache(parsed)
+	if !exists {
```

Comment:
I know this works but I think it's more obvious for the reviewer if you push it into the if statement. At a glance this looks like it might nil `cfg` because that's often what happens when you return an error or not exists. Something like this should work identically but be easier to read:

```
if c, exists := p.getFromCache(parsed); !exists {
```

Reply:
The problem with this is that it doesn't store c so if it gets something in the cache, it won't keep it.